File: desktop-mend/src/main/java/co/samco/mend4/desktop/input/impl/MendInputBox.java
package co.samco.mend4.desktop.input.impl;

import co.samco.mend4.desktop.core.ColorSchemeData;
import co.samco.mend4.desktop.input.InputListener;

import java.awt.Dimension;
import java.awt.Toolkit;
import java.awt.event.KeyEvent;
import java.awt.event.KeyListener;
import java.util.ArrayList;
import java.util.List;

import javax.swing.BorderFactory;
import javax.swing.JFrame;
import javax.swing.JPasswordField;
import javax.swing.JScrollPane;
import javax.swing.JTextArea;
import javax.swing.text.JTextComponent;

public class MendInputBox extends JFrame implements KeyListener {
    private static final long serialVersionUID = -7214084221385969252L;

    private final JTextComponent textInput;
    private List<InputListener> listeners = new ArrayList<InputListener>();

    private boolean shiftDown = false;
    private boolean ctrlDown = false;

    public MendInputBox(boolean decorated, boolean password, int width, int height) {
        setSize(width, height);
        setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        if (!decorated)
            setUndecorated(true);

        Dimension dim = Toolkit.getDefaultToolkit().getScreenSize();
        this.setLocation(dim.width / 2 - this.getSize().width / 2, dim.height / 2 - this.getSize().height / 2);

        if (password) {
            textInput = new JPasswordField();
            textInput.setPreferredSize(new Dimension(width, height));
            this.add(textInput);
        } else {
            JTextArea textArea = new JTextArea();
            textArea.setLineWrap(true);
            textArea.setWrapStyleWord(true);
            textInput = textArea;
            JScrollPane sp = new JScrollPane(textArea);
            sp.setPreferredSize(new Dimension(width, height));
            this.add(sp);
        }

        textInput.setBackground(ColorSchemeData.getColor2());
        textInput.setFont(ColorSchemeData.getConsoleFont());
        textInput.setBorder(BorderFactory.createLineBorder(ColorSchemeData.getColor1()));
        textInput.addKeyListener(this);
    }

    public void addListener(InputListener l) {
        listeners.add(l);
    }

    @Override
    public void keyPressed(KeyEvent e) {
        if (e.getKeyCode() == KeyEvent.VK_SHIFT) {
            shiftDown = true;
        }
        if (e.getKeyCode() == KeyEvent.VK_CONTROL) {
            ctrlDown = true;
        }
    }

    @Override
    public void keyReleased(KeyEvent e) {
        handleControlsUp(e);
        if (e.getKeyCode() == KeyEvent.VK_ENTER) {
            handleSubmit();
        }
    }

    private void handleSubmit() {
        char[] text = getText();

        if (shiftDown || ctrlDown) {
            for (InputListener l : listeners) {
                l.onWrite(text);
            }
        }
        if (shiftDown) {
            close();
        }
        if (ctrlDown) {
            clear();
        }
    }

    private char[] getText() {
        return (textInput instanceof JPasswordField)
                ? ((JPasswordField) textInput).getPassword()
                : textInput.getText().toCharArray();
    }

    private void handleControlsUp(KeyEvent e) {
        if (e.getKeyCode() == KeyEvent.VK_SHIFT) {
            shiftDown = false;
        }
        if (e.getKeyCode() == KeyEvent.VK_CONTROL) {
            ctrlDown = false;
        }
    }

    @Override
    public void keyTyped(KeyEvent e) { }

    public void clear() {
        textInput.setText("");
    }

    public void close() {
        setVisible(false);
        dispose();
    }
}

File: desktop-mend/src/main/java/co/samco/mend4/desktop/input/impl/MendInputBox.kt
package co.samco.mend4.desktop.input.impl

import co.samco.mend4.desktop.core.ColorSchemeData.color1
import co.samco.mend4.desktop.core.ColorSchemeData.color2
import co.samco.mend4.desktop.core.ColorSchemeData.consoleFont
import co.samco.mend4.desktop.input.InputListener
import java.awt.Dimension
import java.awt.Toolkit
import java.awt.event.KeyEvent
import java.awt.event.KeyListener
import java.util.function.Consumer
import javax.swing.BorderFactory
import javax.swing.JFrame
import javax.swing.JScrollPane
import javax.swing.JTextArea
import javax.swing.text.JTextComponent

class MendInputBox(
    width: Int,
    height: Int
) : JFrame(), KeyListener {
    private var textInput: JTextComponent
    private val listeners: MutableList<InputListener> = ArrayList()
    private var shiftDown = false
    private var ctrlDown = false

    init {
        setSize(width, height)

        defaultCloseOperation = EXIT_ON_CLOSE

        val dimens = Toolkit.getDefaultToolkit().screenSize

        this.setLocation(
            dimens.width / 2 - this.size.width / 2,
            dimens.height / 2 - this.size.height / 2
        )

        textInput = JTextArea().apply {
            lineWrap = true
            wrapStyleWord = true
            background = color2
            font = consoleFont
            border = BorderFactory.createLineBorder(color1)
            addKeyListener(this@MendInputBox)
        }

        JScrollPane(textInput).apply {
            preferredSize = Dimension(width, height)
            this@MendInputBox.add(this)
        }
    }

    fun addListener(l: InputListener) {
        listeners.add(l)
    }

    override fun keyPressed(e: KeyEvent) {
        if (e.keyCode == KeyEvent.VK_SHIFT) {
            shiftDown = true
        }
        if (e.keyCode == KeyEvent.VK_CONTROL) {
            ctrlDown = true
        }
    }

    override fun keyReleased(e: KeyEvent) {
        handleControlsUp(e)
        if (e.keyCode == KeyEvent.VK_ENTER) {
            handleSubmit()
        }
    }

    private fun handleSubmit() {
        val text = text
        if (shiftDown || ctrlDown) {
            for (l in listeners) {
                l.onWrite(text)
            }
        }
        if (shiftDown) {
            close()
        }
        if (ctrlDown) {
            clear()
        }
    }

    private val text: CharArray
        get() = textInput.text.toCharArray()

    private fun handleControlsUp(e: KeyEvent) {
        if (e.keyCode == KeyEvent.VK_SHIFT) {
            shiftDown = false
        }
        if (e.keyCode == KeyEvent.VK_CONTROL) {
            ctrlDown = false
        }
    }

    override fun keyTyped(e: KeyEvent) {}

    private fun clear() {
        textInput.text = ""
    }

    private fun close() {
        isVisible = false
        listeners.forEach(Consumer { l: InputListener -> l.onClose() })
        dispose()
    }

    companion object {
        private const val serialVersionUID = -7214084221385969252L
    }
}